ethereum/contractBackend.go

```diff
@@ -0,0 +1,97 @@
+package ethereum
+
+import (
+	"context"
+	"time"
+
+	"github.com/ethereum/go-ethereum/accounts"
+	"github.com/ethereum/go-ethereum/accounts/keystore"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/pkg/errors"
+
+	ethchannel "perun.network/go-perun/backend/ethereum/channel"
+	ethwallet "perun.network/go-perun/backend/ethereum/wallet"
+	"perun.network/go-perun/channel"
+	"perun.network/go-perun/wallet"
+)
+
+// ContractBackend provides ethereum specific contract backend functionality.
+type ContractBackend struct {
+	// cb is the instance of contract backend that will be used for all onchain communications.
+	cb *ethchannel.ContractBackend
+}
+
+// InitCustomContractBackend initializes a contract backend with the pre configured ethereum connection and wallet.
+// Use this function ONLY IN TEST PACKAGES to set a simulated backend and test wallets with weak encryption params.
+func InitCustomContractBackend(cb *ethchannel.ContractBackend) *ContractBackend {
```

Comment:
This would be much better placed in `ethereum/test` or some other test package, otherwise it will be compiled into a release binary.

Reply:
Yeah, this is one thing that i wanted to change, but currently I am not finding a clean way to do this under these constraints:

1. Keep the properties of ContractBackend private, as it is an internal detail.
2. Setup a custom backend for testing, when tests are run in a separate package "ethereum_test" and not in the same package "ethereum". 

One way, i was thinking was to use build tags and move this function to a separate file, that is built  only during tests. But that implies "test" flag should be passed in the go test command everytime. 

Do you find any other better ways ?

Reply:
Looking at it again, this functions could just be called `NewContractBackend` and can then be used to create the `ContractBackend` that is passed to the client.  
The other `InitContractBackend` function can be privately put in the `[ethereum/test]` package. Just need to somehow expose the Scrypt-Params. This resolves the issue afaik.

Reply:
Yes, I have updated the name to `NewXYZ...` for both the backends. No, the function cannot return `*ethchannel.ContractBackend` because if the package is compiled as plugin and the symbol should be loaded, then the code that uses it will have to import `go-perun/ethereum/channel`. This in turn will cause the calling code to depend on `go-ethereum` library (transiently).
I have explained this in detail in the doc comments of the function as well as package docs.

Reply:
Regarding the other point, I figured out a way to do this using the `internal` package feature.

I moved all the implementation code to a separate package within the `internal` directory.  Code from this directory can imported only by packages in the internal's parent directory and its children. So even though the package has exported variables, it is not exposed outside of the ethereum package. However, with this approach I ended up with three packages, on each for real backend, test backend and implementation.

The only other way i saw to just use two packages (real backend and test backend), but that would cause a lot of code duplication. Hence i preferred the previous method.

Reply:
Ok great.